File: golang/005.go
package main

func divide(arr []int) []int {
	//base case
	if len(arr) < 2 {
		return arr
	}

	midpoint := len(arr) / 2
	leftArr := divide(arr[:midpoint])
	rightArr := divide(arr[midpoint:])

	return conquer(leftArr, rightArr)
}

func conquer(leftArr, rightArr []int) []int {
	mainArr := make([]int, len(leftArr)+len(rightArr))
	mainP, leftP, rightP := 0, 0, 0

	for leftP < len(leftArr) && rightP < len(rightArr) {
		if leftArr[leftP] < rightArr[rightP] {
			mainArr[mainP] = leftArr[leftP]
			mainP++
			leftP++
		} else {
			mainArr[mainP] = rightArr[rightP]
			mainP++
			rightP++
		}

	}

	for leftP < len(leftArr) {
		mainArr[mainP] = leftArr[leftP]
		mainP++
		leftP++
	}

	for rightP < len(rightArr) {
		mainArr[mainP] = rightArr[rightP]
		mainP++
		rightP++
	}

	return mainArr
}

func compareArr(arrOne, arrTwo []int) []int {
	sortedArrOne, sortedArrTwo := divide(arrOne), divide(arrTwo)
	pointerOne, pointerTwo := 0, 0
	resArr := []int{}

	if len(sortedArrTwo) < len(sortedArrOne) {
		copyArr := sortedArrTwo
		sortedArrTwo = sortedArrOne
		sortedArrOne = copyArr
	}

	for pointerOne < len(sortedArrOne) && pointerTwo < len(sortedArrTwo) {
		if sortedArrOne[pointerOne] == sortedArrTwo[pointerTwo] {
			if len(resArr)-1 < 0 {
				resArr = append(resArr, sortedArrOne[pointerOne])
				pointerOne++
				pointerTwo++
			} else if sortedArrOne[pointerOne] != resArr[len(resArr)-1] {
				resArr = append(resArr, sortedArrOne[pointerOne])
				pointerOne++
				pointerTwo++
			} else {
				pointerOne++
				pointerTwo++
			}

		} else if sortedArrOne[pointerOne] > sortedArrTwo[pointerTwo] {
			pointerTwo++
		} else if sortedArrOne[pointerOne] < sortedArrTwo[pointerTwo] {
			pointerOne++
		}
	}

	return resArr
}

/*func main() {
	nums1 := []int{4, 5, 6, 7, 3, 23, 5}
	nums2 := []int{2, 5, 6, 2, 4, 4, 2, 8}
	// fmt.Println(divide([]int{3,6,8,3,24,5,5,3,42,4,67,6}))
	fmt.Println(compareArr(nums2, nums1))
}*/

File: python/005.py
def divide(arr):
    if len(arr) < 2:
        return arr
    mid = len(arr) // 2
    leftArr = divide(arr[:mid])
    rightArr = divide(arr[mid:])
    return conquer(leftArr, rightArr)

def conquer(left,right):
    mainArr = list(range(len(left)+len(right)))
    mPointer,lPointer, rPointer = 0,0,0
    while lPointer < len(left) and rPointer < len(right):
        if left[lPointer] < right[rPointer]:
            mainArr[mPointer] = left[lPointer]
            mPointer += 1
            lPointer += 1
        else:
            mainArr[mPointer] = right[rPointer]
            mPointer += 1
            rPointer += 1
    while lPointer < len(left):
        mainArr[mPointer] = left[lPointer]
        mPointer += 1
        lPointer += 1
    while rPointer < len(right):
        mainArr[mPointer] = right[rPointer]
        mPointer += 1
        rPointer += 1
    return mainArr

def findCommonElems(arrOne, arrTwo):
    sortArrOne, sortArrTwo = divide(arrOne), divide(arrTwo)
    onePointer, twoPointer = 0,0
    resultArr = []

    print(len(resultArr))

    if len(sortArrTwo) < len(sortArrOne):
        copyArr = sortArrTwo
        sortArrTwo = sortArrOne
        sortArrOne = copyArr

    while onePointer < len(sortArrOne):
        if sortArrOne[onePointer] == sortArrTwo[twoPointer]:
            if len(resultArr) == 0:
                resultArr.append(sortArrOne[onePointer])
                onePointer += 1
                twoPointer += 1
            elif sortArrOne[onePointer] != resultArr[len(resultArr)-1]:
                resultArr = resultArr.append(sortArrOne[onePointer])
                onePointer += 1
                twoPointer += 1
            else:
                onePointer += 1
                twoPointer += 1
        elif sortArrOne[onePointer] < sortArrTwo[twoPointer]:
            onePointer += 1
        elif sortArrTwo[twoPointer] < sortArrOne[onePointer]:
            twoPointer += 1

    return resultArr

print(divide([6,3,2,8,9,2,5]))
print(findCommonElems([4,5,6,7,3,23,5],[2,5,6,2,4,4,2]))